src/producer/transactionManager.js

```diff
@@ -93,21 +209,111 @@ module.exports = ({ logger, cluster, transactionTimeout = 60000 }) => {
      */
-    getProducerId() {
-      return producerId
+    beginTransaction() {
+      transactionalGuard()
+      stateMachine.transitionTo(STATES.TRANSACTING)
     },
 
     /**
-     * Get the current producer epoch
-     * @returns {number}
+     * Add partitions to a transaction if they are not already marked as participating.
+     *
+     * Should be called prior to sending any messages during a transaction
+     * @param {TopicData[]} topicData
+     *
+     * @typedef {Object} TopicData
+     * @property {string} topic
+     * @property {object[]} partitions
+     * @property {number} partitions[].partition
      */
-    getProducerEpoch() {
-      return producerEpoch
+    async addPartitionsToTransaction(topicData) {
+      transactionalGuard()
+      const newTopicPartitions = {}
+
+      topicData.forEach(({ topic, partitions }) => {
+        transactionTopicPartitions[topic] = transactionTopicPartitions[topic] || {}
+
+        partitions.forEach(({ partition }) => {
+          if (!transactionTopicPartitions[topic][partition]) {
+            newTopicPartitions[topic] = newTopicPartitions[topic] || []
+            newTopicPartitions[topic].push(partition)
+          }
+        })
+      })
+
+      const topics = Object.keys(newTopicPartitions).map(topic => ({
+        topic,
+        partitions: newTopicPartitions[topic],
+      }))
+
+      if (topics.length) {
+        const broker = await findTransactionCoordinator()
+        await broker.addPartitionsToTxn({ transactionalId, producerId, producerEpoch, topics })
+      }
+
+      topics.forEach(({ topic, partitions }) => {
+        partitions.forEach(partition => {
+          transactionTopicPartitions[topic][partition] = true
+        })
+      })
+    },
+
+    /**
+     * Commit the ongoing transaction
+     */
+    async commit() {
+      transactionalGuard()
+      stateMachine.transitionTo(STATES.COMMITTING)
+
+      const broker = await findTransactionCoordinator()
+      await broker.endTxn({ producerId, producerEpoch, transactionalId, transactionalResult: true })
+
+      stateMachine.transitionTo(STATES.READY)
+    },
+
+    /**
+     * Abort the ongoing transaction
+     */
+    async abort() {
+      transactionalGuard()
+      stateMachine.transitionTo(STATES.ABORTING)
+
+      const broker = await findTransactionCoordinator()
+      await broker.endTxn({
+        producerId,
+        producerEpoch,
+        transactionalId,
+        transactionalResult: false,
+      })
+
+      stateMachine.transitionTo(STATES.READY)
+    },
+
+    /**
+     * Whether the producer id has already been initialized
+     */
+    isInitialized() {
+      return producerId !== NO_PRODUCER_ID
+    },
+
+    isTransactional() {
+      return transactional
+    },
+
+    isInTransaction() {
+      return currentState === STATES.TRANSACTING
     },
   }
 
+  // Enforce the state machine
+  stateMachine.guard(transactionManager, 'initProducerId', [STATES.UNINITIALIZED, STATES.READY])
```

Comment:
Since we reference the state machine within these methods to transition state I think it'd be a bit weird to add the guards during instantiation. However I could add a method to add all the guard on instantiation of the transaction manager, which should provide the clarity you're asking for.